### PR TITLE
migrate e2e & integration role_test to common

### LINKED ISSUE #13637: Unify testing framework
Test flakiness and maintenance cost remains one of the larger issues for Etcd project. Proposals like https://github.com/etcd-io/etcd/issues/13167 to track flakiness and contributions to address singular tests have definitely helped however to fully resolve the problem we also need to address the root cause. Etcd projects currently maintains multiple disconnected ways to run tests (unit, integration, e2e, functional) that mostly verify the same scenarios. Testing features on different level is itself desired as it allows to  isolate failures and speed up resolution, however not in a way that is currently done in Etcd. Almost zero test scenario and test framework is reused. This means that we have multiplied test code lines without any benefit.

Goal:
* Reduce number of test code without affecting test coverage
* Simplify process of adding new integration/e2e test
* Make it easier to isolate test failures by comparing test results from different methods

I propose to unify testing by identifying common test scenarios, test framework operations and making them accessable the same way no matter which test method is used (unit, integration, e2e, functional). For example simple put&get test scenario goes through very same stages no matter if this is integration or e2e cluster: create cluster, execute put, execute get, compare results, cleanup. This test should be written once and executed on different layers, no matter if underneath framework runs a fake grpc client or starts whole process. With this we will reduce number of code lines as each test will be needed to be implemented once, hide all the differences between test types reducing knowledge needed to add new tests and improve deflaking process by comparing same test results from different methods (we can isolate test location, if both e2e and integration tests fail then feature is broken, if only one of them then the test framework)

At this moment I would be most effective to target unifying integration and e2e tests as they are most similar in both cluster setup and test scenarios. We should be able to identify minimal common interface for cluster creation and communication that would allow us to start rewriting tests. 

Plan:
* Create new `Cluster` and `Client` interfaces in `tests/framework`.
* Find common function in `tests/framework/e2e/cluster.go` and `tests/framework/integration/cluster.go` and migrate them one by one.
* Deduplicate test scenarios between e2e and integration tests
* Create common tests located in `tests/common` that will be invoked by e2e and integration.

cc @ptabor @ahrtr @spzala

### PR DESCRIPTION
**Pull request by chaochn47**
Partially fix #13637

### CODE REVIEW

File: tests/e2e/ctl_v3_role_test.go [hunk 1]
```diff
@@ -21,17 +21,8 @@ import (
...
 )
 
-func TestCtlV3RoleAdd(t *testing.T)      { testCtl(t, roleAddTest) }
-func TestCtlV3RootRoleGet(t *testing.T)  { testCtl(t, rootRoleGetTest) }
-func TestCtlV3RoleAddNoTLS(t *testing.T) { testCtl(t, roleAddTest, withCfg(*e2e.NewConfigNoTLS())) }
-func TestCtlV3RoleAddClientTLS(t *testing.T) {
-	testCtl(t, roleAddTest, withCfg(*e2e.NewConfigClientTLS()))
-}
-func TestCtlV3RoleAddPeerTLS(t *testing.T) { testCtl(t, roleAddTest, withCfg(*e2e.NewConfigPeerTLS())) }
 func TestCtlV3RoleAddTimeout(t *testing.T) { testCtl(t, roleAddTest, withDialTimeout(0)) }
```
Comment (chaochn47) [inline tests/e2e/ctl_v3_role_test.go:24]: I don't think we can easily make the client dial-timeout configurable both in e2e and integration tests. So I leave it alone.

File: tests/e2e/ctl_v3_role_test.go [hunk 2]
```diff
@@ -21,17 +21,8 @@ import (
...
 
-func TestCtlV3RoleGrant(t *testing.T) { testCtl(t, roleGrantTest) }
-
 func roleAddTest(cx ctlCtx) {
```
Comment (chaochn47) [inline tests/e2e/ctl_v3_role_test.go:26]: Called from `TestCtlV3RoleAddTimeout`

File: tests/e2e/ctl_v3_role_test.go [hunk 3]
```diff
@@ -58,94 +49,6 @@ func roleAddTest(cx ctlCtx) {
...
-	}
-}
-
-func ctlV3RoleMultiExpect(cx ctlCtx, args []string, expStr ...string) error {
-	cmdArgs := append(cx.PrefixArgs(), "role")
-	cmdArgs = append(cmdArgs, args...)
-
-	return e2e.SpawnWithExpects(cmdArgs, cx.envMap, expStr...)
-}
 func ctlV3Role(cx ctlCtx, args []string, expStr string) error {
```
Comment (chaochn47) [inline tests/e2e/ctl_v3_role_test.go:52]: Line 52 and below are retained only in this PR. After auth_test are migrated to common, they should be cleaned up.

File: tests/e2e/ctl_v3_role_test.go (hunk 1 above)
Comment (ahrtr) [reply tests/e2e/ctl_v3_role_test.go:24]: Please add a comment here.

File: tests/common/role_test.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+			testutils.ExecuteWithTimeout(t, 10*time.Second, func() {
+				_, err := cc.RoleAdd("root")
+				if err != nil {
+					t.Fatalf("want no error, but got (%v)", err)
+				}
+				_, err = cc.RoleAdd("root")
+				if err == nil || !strings.Contains(err.Error(), rpctypes.ErrRoleAlreadyExist.Error()) {
+					t.Fatalf("want (%v) error, but got (%v)", rpctypes.ErrRoleAlreadyExist, err)
+				}
+				_, err = cc.RoleAdd("")
```
Comment (serathius) [inline tests/common/role_test.go:70]: This looks like a separate test, like `TestRoleError` was merged with other test.

**Comment by chaochn47** [comment]
All comments are addressed. PTAL, thx! @ahrtr @serathius

File: client/v3/maintenance.go [hunk 5]
```diff
@@ -181,13 +181,16 @@ func (m *maintenance) AlarmDisarm(ctx context.Context, am *AlarmMember) (*AlarmR
...
+		header := ar.Header
 		for _, am := range ar.Alarms {
 			dresp, derr := m.AlarmDisarm(ctx, (*AlarmMember)(am))
 			if derr != nil {
 				return nil, toErr(ctx, derr)
 			}
 			ret.Alarms = append(ret.Alarms, dresp.Alarms...)
+			header = dresp.Header
 		}
+		ret.Header = header
```
Comment (serathius) [inline client/v3/maintenance.go:193]:
```suggestion
		ret.Header = ar.Header
		for _, am := range ar.Alarms {
			dresp, derr := m.AlarmDisarm(ctx, (*AlarmMember)(am))
			if derr != nil {
				return nil, toErr(ctx, derr)
			}
			ret.Alarms = append(ret.Alarms, dresp.Alarms...)
			ret.Header = dresp.Header
		}
```

Comment (serathius) [reply client/v3/maintenance.go:193]: This looks like a reasonable fix for a bug, however I would want to make sure that it's properly reviewed. Can you send it in separate PR?

File: tests/common/user_test.go [hunk 6]
```diff
@@ -301,7 +302,8 @@ func TestUserDelete(t *testing.T) {
 				if err == nil {
 					t.Fatalf("deleting a non-existent user should fail")
 				}
-				assert.Contains(t, err.Error(), "user name not found")
+				expectErrMsg := "user name not found"
```
Comment (serathius) [inline tests/common/user_test.go:305]: Looks like unrelated change.

File: tests/framework/e2e/etcd_spawn.go [hunk 7]
```diff
@@ -15,10 +15,27 @@
...
+
 	"go.etcd.io/etcd/pkg/v3/expect"
 	"go.uber.org/zap"
 )
 
 func SpawnCmd(args []string, envVars map[string]string) (*expect.ExpectProcess, error) {
 	return SpawnCmdWithLogger(zap.NewNop(), args, envVars)
 }
+
+// SpawnCmdWithJsonPrinter spawns an etcdctl process with provided command line arguments and environment variables
```
Comment (serathius) [inline tests/framework/e2e/etcd_spawn.go:28]: This function has strong assumptions about working only for `etcdctl`, this makes both name and location (`etcd_spawn.go`) not match expectations. Please consider making this a helper method for `EtcdctlV3` struct below.

File: tests/framework/e2e/etcd_spawn.go [hunk 8]
```diff
@@ -15,10 +15,27 @@
...
+func SpawnCmdWithJsonPrinter(args []string, envVars map[string]string) (string, error) {
+	args = append(args, "-w", "json")
+	cmd, err := SpawnCmd(args, envVars)
+	if err != nil {
+		return "", err
+	}
+	return cmd.ExpectFunc(func(txt string) bool {
+		// lease response does not have header substring
+		// member add response does not have revision substring
+		return strings.Contains(txt, "header") || strings.Contains(txt, "revision")
```
Comment (serathius) [inline tests/framework/e2e/etcd_spawn.go:39]: I consider using or here a workaround for tests, however I think more important here is to find out why we have such inconsistency and consider fixing it. Could you give exact examples of `Response` protos in https://github.com/etcd-io/etcd/blob/main/api/etcdserverpb/rpc.proto that are missing some fields?

File: tests/framework/e2e/etcdctl.go [hunk 9]
```diff
@@ -120,8 +121,7 @@ func (ctl *EtcdctlV3) Put(key, value string, opts config.PutOptions) error {
 }
 
 func (ctl *EtcdctlV3) Delete(key string, o config.DeleteOptions) (*clientv3.DeleteResponse, error) {
-	args := ctl.cmdArgs()
-	args = append(args, "del", key, "-w", "json")
+	args := ctl.cmdArgs("del", key)
```
Comment (serathius) [inline tests/framework/e2e/etcdctl.go:124]: Thanks for refactor, however I'm worried that this might make a breaking change for othere when they migrate. Could we move it to separate PR so it can be quickly reviewed and merged?

File: tests/framework/e2e/etcdctl.go [hunk 10]
```diff
@@ -507,3 +508,100 @@ func (ctl *EtcdctlV3) UserChangePass(user, newPass string) error {
...
+	}
+	var resp clientv3.AuthRoleAddResponse
+	line, err := cmd.Expect("header")
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal([]byte(line), &resp)
+	return &resp, err
+}
+
```
[the remark this answers is not in the document]
Comment (serathius) [reply tests/framework/e2e/etcdctl.go:449]: It's a good idea, however let's try not to mix in to many changes into PR.

File: tests/framework/integration.go [hunk 11]
```diff
@@ -44,8 +44,8 @@ func (e integrationRunner) NewCluster(t testing.TB, cfg config.ClusterConfig) Cl
 	var err error
 	var integrationCfg integration.ClusterConfig
 	integrationCfg.Size = cfg.ClusterSize
-	integrationCfg.ClientTLS, err = tlsInfo(t, cfg.ClientTLS)
 	integrationCfg.QuotaBackendBytes = cfg.QuotaBackendBytes
+	integrationCfg.ClientTLS, err = tlsInfo(t, cfg.ClientTLS)
```
Comment (serathius) [inline tests/framework/integration.go:48]: Not sure this is related.

File: tests/framework/interface.go [hunk 12]
```diff
@@ -39,6 +39,7 @@ type Member interface {
 	Stop()
 }
 
+// Client is a wrapper on top of "go.etcd.io/etcd/client/v3".Client interface
```
Comment (serathius) [inline tests/framework/interface.go:42]: Not sure what what `wrapper on top of ... interface` means. I would set that this interface is meant as a subset of `clientv3.Client` interface implemented by common framework, however this is also not true as options are passed differently to be more expressive. 

Overall this is still work in progress, so I would propose that you skip adding documentation in this PR.

File: tests/framework/e2e/etcdctl.go (hunk 9 above)
Comment (chaochn47) [reply tests/framework/e2e/etcdctl.go:124]: Make sense.

File: tests/framework/e2e/etcd_spawn.go (hunk 8 above)
Comment (chaochn47) [reply tests/framework/e2e/etcd_spawn.go:39]: Yeah, let me raise a issue instead of fixing it in the tests migration PR.

File: tests/common/user_test.go (hunk 6 above)
Comment (chaochn47) [reply tests/common/user_test.go:305]: Not related.

**Comment by chaochn47** [comment]
Thanks for the review. I will keep the current PR for the minimal changes of `role_test` test migration. 

The issues/bugs exposed by refactoring with etcdctl json printer should be addressed in a separate PR.

File: tests/framework/e2e/etcd_spawn.go (hunk 8 above)
Comment (chaochn47) [reply tests/framework/e2e/etcd_spawn.go:39]: It's not grpc client implementation missed some fields but client v3. 

Examples:
1. `header` substring is missing because of embedding.
https://github.com/etcd-io/etcd/blob/b7be91f98b7a7a6202ed9e9bc37b692d484585a8/client/v3/lease.go#L71-L75
2. `revision` is not populated in
https://github.com/etcd-io/etcd/blob/b7be91f98b7a7a6202ed9e9bc37b692d484585a8/server/etcdserver/api/v3rpc/member.go#L108-L110

Comment (ahrtr) [reply tests/framework/e2e/etcd_spawn.go:39]: Thanks for raising this. 

For lease, we need some big change on this. There is an [issue](https://github.com/etcd-io/etcd/issues/13915) discussing this.  The problem mentioned here is lease response indeed has the header info, but it isn't included under the "header" section. For example,

```
$ etcdctl lease list -w json | jq
{
  "cluster_id": 14841639068965180000,
  "member_id": 10276657743932975000,
  "revision": 1,
  "raft_term": 2,
  "leases": []
}

$ etcdctl  get k1 -w json | jq
{
  "header": {
    "cluster_id": 14841639068965180000,
    "member_id": 10276657743932975000,
    "revision": 1,
    "raft_term": 2
  }
}
```

With regard to the missing "revision" in the memer related commands, I think we might need to fix it. Feel free to raise a PR to send out for review. 
```
$ etcdctl member list  -w json | jq
{
  "header": {
    "cluster_id": 14841639068965180000,
    "member_id": 10276657743932975000,
    "raft_term": 2
  },
  "members": [
    {
      "ID": 10276657743932975000,
      "name": "default",
      "peerURLs": [
        "http://localhost:2380"
      ],
      "clientURLs": [
        "http://localhost:2379"
      ]
    }
  ]
}
```

Comment (chaochn47) [reply tests/framework/e2e/etcd_spawn.go]: > The problem mentioned here is lease response indeed has the header info, but it isn't included under the "header" section

Exactly. I think may be a json tag added to the LeaseLeasesResponse struct should help. It's a relatively small change.

Comment (ahrtr) [reply tests/framework/e2e/etcd_spawn.go]: Technically speaking, it's a small change. But it has impact on user experience. So we should be cautious about this.  Let's track this in the issue I mentioned above.

Feel free to fix the mission revision in member related commands for now.

Comment (chaochn47) [reply tests/framework/e2e/etcd_spawn.go]: Yeah, 100% agree. I just realized it like kube-apiserver will be impacted by this small change so I dropped the drafted PR.

File: tests/framework/e2e/etcdctl.go [hunk 13]
```diff
@@ -507,3 +508,79 @@ func (ctl *EtcdctlV3) UserChangePass(user, newPass string) error {
...
+// SpawnCmdWithJsonPrinter spawns an etcdctl process with provided command line arguments and environment variables
+// with json printer output and capture header line as output.
+func SpawnCmdWithJsonPrinter(args []string, envVars map[string]string) (string, error) {
+	args = append(args, "-w", "json")
+	cmd, err := SpawnCmd(args, envVars)
+	if err != nil {
+		return "", err
+	}
+	return cmd.ExpectFunc(func(txt string) bool {
+		// header substring is missing because of embedding in LeaseLeasesResponse.
```
Comment (serathius) [inline tests/framework/e2e/etcdctl.go:582]: This PR doesn't touch `LeaseLeasesResponse`, let's skip this for now.

File: tests/framework/e2e/etcdctl.go [hunk 14]
```diff
@@ -507,3 +508,79 @@ func (ctl *EtcdctlV3) UserChangePass(user, newPass string) error {
...
+		// revision substring is missing because ClusterServer.header() does not fill revision in the header
+		return strings.Contains(txt, "header") || strings.Contains(txt, "revision")
+	})
+}
```
Comment (serathius) [inline tests/framework/e2e/etcdctl.go:586]:
```suggestion
func (ctl *EtcdctlV3) RoleDelete(role string) (*clientv3.AuthRoleDeleteResponse, error) {
    var resp clientv3.AuthRoleDeleteResponse
    err := ctl.spawnJsonCmd(&resp, "role", "delete",  role)
    return &resp, err
}

func (ctl *EtcdctlV3) spawnJsonCmd(output interface{}, args ...string) error {
	args = append(args, "-w", "json")
	cmd, err := SpawnCmd(append(ctl.cmdArgs(), args...), nil)
	if err != nil {
		return err
	}
	line, err := cmd.Expect("header")
	if err != nil {
		return err
	}
	err = json.Unmarshal([]byte(line), output)
	if err != nil {
		return err
	}
	return nil
}
```

`SpawnCmdWithJsonPrinter` you proposed doesn't work for any cmd, it depends directly on `-w json` parameters available only for some of the etcdctl commands. As so it should not be a function, but a method on EtcdctlV3 struct. 

I also proposed to make one additional step to include json parsing. It make sense that if we provide commands used to force json output, then we should always be able to parse json.

**Review by serathius — APPROVED** [review]
